### PR TITLE
support for RFC9380 DST G1 scheme

### PR DESCRIPTION
**Pull request by CluEleSsUK**
- refactored the IBE code to be more generalised

### CODE REVIEW

**Comment by CluEleSsUK** [comment]
Created a draft of this - just realised I haven't updated the client yet, so should probably do that first, then update tlock go, then create test vectors

File: test/crypto/rfc-scheme.test.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+describe("rfc compliant scheme", () => {
+    it("should pass test vectors from go codebase", () => {
+
```
Comment (AnomalRoil) [inline test/crypto/rfc-scheme.test.ts:3]: Feels like this one is missing the tests, no?

**Review by AnomalRoil — COMMENTED** [review]
Overall LGTM, but these are very breaking changes for the  users, since the API of the function themselves changed, no?

File: test/crypto/rfc-scheme.test.ts (hunk 1 above)
Comment (CluEleSsUK) [reply test/crypto/rfc-scheme.test.ts:18]: oops, have added it

**Comment by CluEleSsUK** [comment]
yes and no - the interface for encryption and decryption hasn't changed, though users who have been using `MAINNET_URL` or `mainnetClient()` who upgrade will now automagically use `quicknet` rather than `fastnet`. In a sense this is a breaking change.
I'll bump the major version

**Comment by CluEleSsUK** [comment]
actually wait it's still in `0.x.x`, so I think it's fine to bump minor

File: src/crypto/ibe.ts [hunk 2]
```diff
@@ -50,91 +56,87 @@ export async function encryptOnG1(master: PointG1, ID: Uint8Array, msg: Uint8Arr
...
+        (id: Uint8Array) => bls.PointG2.hashToCurve(id),
+        (m, Qid) => bls.pairing(m, Qid)
+    )
+}
 
-    // 6. Compute M XOR H(sigma)
-    const hsigma = h4(sigma, msg.length)
+export async function encryptOnG2(master: PointG2, ID: Uint8Array, msg: Uint8Array): Promise<Ciphertext<PointG2>> {
+    return encrypt(master, ID, msg, PointG2.BASE,
+        (id: Uint8Array) => bls.PointG1.hashToCurve(id),
```
Comment (AnomalRoil) [inline src/crypto/ibe.ts:68]: Could you explicitly specify the DST here?

File: src/index.ts [hunk 3]
```diff
@@ -40,21 +47,40 @@ export async function timelockDecrypt(
...
+    return new HttpChainClient(chain, opts, {
         userAgent: `tlock-js-${LIB_VERSION}`
     })
 }
 
 export function mainnetClient(): HttpChainClient {
+    const opts = {
+        ...defaultChainOptions,
+        chainVerificationParams: {
+            chainHash: "52db9ba70e0cc0f6eaf7803dd07447a1f5477735fd3f661792ba94600c84e971",
```
Comment (AnomalRoil) [inline src/index.ts:67]: Could have added a comment with the name of the network :P

File: src/index.ts [hunk 4]
```diff
@@ -40,21 +47,40 @@ export async function timelockDecrypt(
...
+    }
+    const chain = new HttpCachingChain(MAINNET_CHAIN_URL, opts)
+    return new HttpChainClient(chain, opts)
+}
+
+export function nonRFCMainnetClient(): HttpChainClient {
     const opts = {
         ...defaultChainOptions,
         chainVerificationParams: {
             chainHash: "dbd506d6ef76e5f386f41c651dcb808c5bcbd75471cc4eafa3f4df7ad4e4c493",
```
Comment (AnomalRoil) [inline src/index.ts:79]: Could have added a comment with the name of the network :P

**Review by AnomalRoil — APPROVED** [review]
LGTM

**Pull request merged**